ddtrace/context.py

```diff
@@ -1,180 +1,100 @@
 from .constants import SAMPLING_PRIORITY_KEY
 from .internal.logger import get_logger
-from .utils.formats import asbool
-from .utils.formats import get_env
 
 
 if TYPE_CHECKING:
-    from ddtrace import Span
+    from .internal.compat import NumericType
+    from .span import Span
+    from .span import _MetaDictType
+    from .span import _MetricDictType
 
 log = get_logger(__name__)
 
 
 class Context(object):
-    """
-    Context is used to keep track of a hierarchy of spans for the current
-    execution flow. During each logical execution, the same ``Context`` is
-    used to represent a single logical trace, even if the trace is built
-    asynchronously.
-
-    A single code execution may use multiple ``Context`` if part of the execution
-    must not be related to the current tracing. As example, a delayed job may
-    compose a standalone trace instead of being related to the same trace that
-    generates the job itself. On the other hand, if it's part of the same
-    ``Context``, it will be related to the original trace.
-
-    This data structure is thread-safe.
+    """Represents a snapshot of a trace to be used to propagate a trace
+    across execution boundaries (eg. distributed tracing).
+
+    A Context contains the span_id of the active span at the time the context
+    is created.
     """
 
-    _partial_flush_enabled = asbool(get_env("tracer", "partial_flush_enabled", default=False))
-    _partial_flush_min_spans = int(get_env("tracer", "partial_flush_min_spans", default=500))  # type: ignore[arg-type]
+    __slots__ = ["_trace_id", "_span_id", "_span", "_meta", "_metrics"]
 
     def __init__(
         self,
         trace_id=None,  # type: Optional[int]
         span_id=None,  # type: Optional[int]
-        sampling_priority=None,  # type: Optional[int]
+        sampling_priority=None,  # type: Optional[NumericType]
         dd_origin=None,  # type: Optional[str]
     ):
         # type: (...) -> None
-        """
-        Initialize a new thread-safe ``Context``.
-
-        :param int trace_id: trace_id of parent span
-        :param int span_id: span_id of parent span
-        """
-        self._current_span = None  # type: Optional[Span]
-        self._lock = threading.Lock()
-
-        self._parent_trace_id = trace_id
-        self._parent_span_id = span_id
-        self._sampling_priority = sampling_priority
-        self._local_root_span = None  # type: Optional[Span]
+        self._trace_id = trace_id
+        self._span_id = span_id
+
+        # TODO[v1.0]: we need to keep a reference back to the span to maintain
+        # backwards compatibility when using context as a parent.
+        # eg:
+        #    ctx = span.context
+        #    tracer.start_span(child_of=ctx)
+        self._span = None  # type: Optional[Span]
+        self._meta = {}  # type: _MetaDictType
+        self._metrics = {}  # type: _MetricDictType
         self.dd_origin = dd_origin
+        self.sampling_priority = sampling_priority
```

Comment:
Hmm I gave it a shot but ran into issues with `dd_origin` and `sampling_priority`. attr won't include them in the initializer because they're properties 😕   

```python
import attr

@attr.s
class A(object):
    c = attr.ib(default=1)
    _d = attr.ib(factory=dict)
    @property
    def c(self):
            return self._d["c"]
    @c.setter
    def c(self, val):
            self._d["c"] = val

>>> A()
A(_d={})
>>> A(c=3)
TypeError: __init__() got an unexpected keyword argument 'c'
```

Reply:
Here's the hack:

```python
import attr

@attr.s
class A(object):
    _c = attr.ib(default=1)
    _d = attr.ib(factory=dict)

    def __attrs_post_init__(self):
        self.c = self._c

    @property
    def c(self):
            return self._d["c"]

    @c.setter
    def c(self, val):
            self._d["c"] = val

a = A(c=3)
print(a.c) # 3
print(a._d) # {'c': 3}

a = A()
print(a.c) # 1
print(a._d) # {'c': 1}
```

When you use a private (prefixed with `_`) attribute with `attr`, it still uses the non-private name (without the `_`) as kwarg for `__init__`. That make sure you can use the public var name but override with a `property`.

Reply:
Hmm the drawback here is that we'll have an attribute (`_c`) defined on every instance of this class which is used only to initialize the value in `_d` which seems like a waste given that there could be many context objects. That along with the `__attrs_post_init__` required I'm not if we gain much here from using attr. WDYT?